Develop > Interoperability > Send Messages (#187)
polkadot-developers/polkadot-docs: develop/interoperability/send-messages.md

---

> title: XCM - Send Messages

Reviewer: Maybe for SEO purposes, this could be better; idk
@eshaben or @dawnkelly09 can definitely help with this comment for sure

```suggestion
title: Send XCM Messages
```

Reviewer 2: Agree with Nico here

description: Unlock blockchain interoperability with XCM — Polkadot's Cross-Consensus Messaging format for cross-chain interactions.
---

> # XCM - Send Messages

Reviewer 2: Use the same name that you put in the title section

## XCM Frame Pallet Overview

> The XCM pallet [(pallet-xcm)](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/pallet-xcm/src/lib.rs) provides a set of pre-defined, commonly used XCVM programs in the form of a set of extrinsics using [FRAME](https://docs.substrate.io/reference/frame-pallets/).

Reviewer: WDYT about adding the following links? just to complement the documentation here

```suggestion
The XCM pallet ([`pallet-xcm`](https://paritytech.github.io/polkadot-sdk/master/pallet_xcm/index.html){target=\_blank}) provides a set of pre-defined, commonly used [`XCVM programs`](https://github.com/polkadot-fellows/xcm-format?tab=readme-ov-file#12-the-xcvm){target=\_blank} in the form of a set of extrinsics using [FRAME](https://docs.substrate.io/reference/frame-pallets/).
```

> This pallet provides some default implementations for traits required by XcmConfig. The XCM executor is also included as an associated type within the pallet's configuration. For further details on XCM Configurations, see the previous page here.

Reviewer: Some feedback here:

- I'm checking the docs of `pallet-xcm` and I struggling to find this: "provides some default implementations for traits". Are you referring to https://paritytech.github.io/polkadot-sdk/master/pallet_xcm/index.html#functions ?

- I think we should add a reference to `XcmConfig` coming from polkadot sdk rust docs

- XCM executor should be linked to https://paritytech.github.io/polkadot-sdk/master/staging_xcm_executor/struct.XcmExecutor.html

> Where the XCM format defines a set of instructions used to construct XCVM programs, pallet-xcm defines a set of extrinsics that can be utilized to build XCVM programs, either to target the local or external chains. pallet-xcm's functionality is separated into three categories:

Reviewer: ```suggestion
Where the [`XCM format`](https://github.com/polkadot-fellows/xcm-format){target=\_blank} defines a set of instructions used to construct XCVM programs, `pallet-xcm` defines a set of extrinsics that can be utilized to build XCVM programs, either to target the local or external chains. The `pallet-xcm` functionality is divided into three categories:
```

> **Primitive** - Dispatchable functions to locally execute an XCM
> 
> **High-level** - Dispatchable functions for asset transfers
> 
> **Version negotiation-specific** - Dispatchable functions

Reviewer: just formatting stuff:

```suggestion
**Primitive** - dispatchable functions to locally execute an XCM

- **High-level** - dispatchable functions for asset transfers

- **Version negotiation-specific** - dispatchable functions
```

Reviewer: is there any link to support these categories? I think it'd be helpful for those trying to get a high-level overview of the pallet-xcm. Or maybe, we can expand this descriptions just a bit to improve the story telling here

This page will highlight the two Primary Primitive Extrinsics which are responsible for sending and executing XCVM programmes as dispatchable functions within the pallet.

> 

Reviewer: maybe we can complement this intro with something at a high level about the roles of pallet-xcm, just like: https://polkadot-blockchain-academy.github.io/pba-book/xcm/pallet/page.html#key-roles-of-pallet-xcm

## Two Primary Extrinsics

### Execute

This call contains direct access to the XCM executor. It is the job of the executor to check the message and ensure that no barrier/filter will block the execution of the XCM.

Once it is deemed valid, the message will then be locally executed, therein returning the outcome as an event. This operation is executed on behalf of whichever account has signed the extrinsic.

> **Note:** It is possible for partial executions to occur.

Reviewer 2: ```suggestion
!!!warning
    It is possible for partial executions to occur.
```

### Send

This call specifies where a message should be sent (via a transport method) externally to a particular destination, i.e. a parachain, smart contract, or any system which is governed by consensus. In contrast to execute, the executor is not called locally, as the execution will occur on the destination chain.

**XCM Router**

It’s important to note that the XCM pallet needs the XcmRouter to send XCMs. It is used to dictate where XCms are allowed to be sent, and which XCM transport protocol to use. 
E.G. Kusama uses the ChildParachainRouter which only allows for Downward Message Passing from the relay to parachains to occur.

For more information about XCM transport protocols, check out the XCM Channels page here.

> 

Reviewer: For both methods, I think it'd be helpful to add links to the original source (polkadot sdk rust docs), same for XCM Router

---
title: Send XCM Messages
description: Unlock blockchain interoperability with XCM — Polkadot's Cross-Consensus Messaging format for cross-chain interactions.
---

> # XCM - Send Messages

Reviewer 3: ```suggestion
# Send XCM Messages
```

!!!note
    For further details on the XCM configuration, refer to the [XCM Configuration](/develop/interoperability/xcm-config/) page.

Where the [`XCM format`](https://github.com/polkadot-fellows/xcm-format){target=\_blank} defines a set of instructions used to construct XCVM programs, `pallet-xcm` defines a set of extrinsics that can be utilized to build XCVM programs, either to target the local or external chains. The `pallet-xcm` functionality is divided into three categories:

- **Primitive** - dispatchable functions to execute XCM locally

- **High-level** - functions for asset transfers between chains

- **Version negotiation-specific** - functions for managing XCM version compability

### Key Roles of `pallet-xcm`

The XCM pallet plays a central role in managing cross-chain messages, with its primary responsibilities including:

> - **Execute XCM Messages** - interacts with the XCM executor to validate and execute messages, adhering to predefined security and filter criteria
> 
> - **Send Messages Across Chains** - allows authorized origins to send XCM messages, enabling controlled cross-chain communication
> 
> - **Reserve-Based Transfers and Teleports** - supports asset movement between chains, governed by filters that restrict operations to authorized origins
> 
> - **XCM Version Negotiation** - ensures compatibility by selecting the appropriate XCM version for inter-chain communication
> 
> - **Asset Trapping and Recovery** - manages trapped assets, enabling safe reallocation or recovery when issues occur during cross-chain transfers
> 
> - **Support for XCVM Operations** - oversees state and configuration requirements necessary for executing cross-consensus programs within the XCVM framework

Reviewer 2: ```suggestion
- **Execute XCM messages** - interacts with the XCM executor to validate and execute messages, adhering to predefined security and filter criteria

- **Send messages across chains** - allows authorized origins to send XCM messages, enabling controlled cross-chain communication

- **Reserve-based transfers and teleports** - supports asset movement between chains, governed by filters that restrict operations to authorized origins

- **XCM version negotiation** - ensures compatibility by selecting the appropriate XCM version for inter-chain communication

- **Asset trapping and recovery** - manages trapped assets, enabling safe reallocation or recovery when issues occur during cross-chain transfers

- **Support for XCVM operations** - oversees state and configuration requirements necessary for executing cross-consensus programs within the XCVM framework
```

> ## Primary Calls of XCM Pallet

Reviewer 2: Maybe replace `Calls` with `Extrinsics` ?

This page will highlight the two **Primary Primitive Calls** which are responsible for sending and executing XCVM programmes as dispatchable functions within the pallet.

### Execute

The [`execute`](https://paritytech.github.io/polkadot-sdk/master/pallet_xcm/pallet/enum.Call.html#variant.execute){target=\_blank} call provides direct interaction with the XCM executor, allowing for the execution of XCM messages originating from a locally signed origin. The executor validates the message, ensuring it complies with any configured barriers or filters before proceeding with execution.

Once validated, the message is executed locally, and an event is emitted to indicate the result—whether the message was fully executed or only partially completed. Execution is capped by a maximum weight ([`max_weight`](https://paritytech.github.io/polkadot-sdk/master/pallet_xcm/pallet/enum.Call.html#variant.execute.field.max_weight){target=\_blank}), and if the required weight exceeds this limit, the message will not be executed.

```rust
pub fn execute<T: Config>(
    message: Box<VersionedXcm<<T as Config>::RuntimeCall>>,
    max_weight: Weight,
)
```

> !!!note

Reviewer 2: The link in the note is not there

    For further details on the Execute extrinsic, see the pallet-xcm documentation{target=_blank}.

!!!warning
    Partial execution of messages may occur depending on the constraints or barriers applied.

<!-- TODO: we should complement this page with some real examples of where the .execute() call is used, or maybe how to use it through papi or something like that -->

### Send

The [send](https://paritytech.github.io/polkadot-sdk/master/pallet_xcm/pallet/enum.Call.html#variant.send){target=\_blank} call enables sending XCM messages to a specified destination. This could be a parachain, smart contract, or any external system governed by consensus. Unlike the execute call, the message is not executed locally; instead, it is transported to the destination chain for processing.

The destination is defined using a [Location](https://paritytech.github.io/polkadot-sdk/master/xcm_docs/glossary/index.html#location){target=\_blank}, which describes the target chain or system. This ensures precise delivery through the configured XCM transport mechanism.

```rust
pub fn send<T: Config>(
    dest: Box<MultiLocation>,
    message: Box<VersionedXcm<<T as Config>::RuntimeCall>>,
)
```

> !!!note For further details on the Send extrinsic, refer to the pallet-xcm documentation{target=_blank}.

Reviewer 2: Fix note formatting and add link
